### PR TITLE
Updated httpmanager to support mtls

### PR DESCRIPTION
**Pull request by gladjohn**
- Updated httpmanager to support mtls
- Adds Net472 support 
-

Tests : Unit and Integration

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 1]
```diff
@@ -167,20 +88,45 @@ private static HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary
...
             {
                 if (cancellationToken.IsCancellationRequested)
                 {
-                    logger.Info("The HTTP request was cancelled. ");
+                    logger.Info("The HTTP request was canceled. ");
                     throw;
                 }
 
                 logger.Error("The HTTP request failed. " + exception.Message);
+                isRetriable = true;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:103]: Why retry on task cancelled exception? Isn't this cancellation from the user?

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 2]
```diff
@@ -189,7 +135,7 @@ private static HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary
             }
 
             // package 500 errors in a "service not available" exception
-            if (IsRetryableStatusCode((int)response.StatusCode))
+            if ((int)response.StatusCode >= 500 && (int)response.StatusCode < 600)
             {
                 throw MsalServiceExceptionFactory.FromHttpResponse(
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:140]: Can you please add the status code in the error message and the uri that produced it? (just host and path, not full uri)

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 3]
```diff
@@ -200,11 +146,39 @@ private static HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary
...
-        protected async Task<HttpResponse> ExecuteAsync(
+        private HttpClient GetHttpClient(X509Certificate2 x509Certificate2)
+        {
+            if (_httpClientFactory is IMsalMtlsHttpClientFactory msalMtlsHttpClientFactory)
+            {
+                // If the factory is an IMsalMtlsHttpClientFactory, use it to get an HttpClient with the certificate
+                return msalMtlsHttpClientFactory.GetHttpClient(x509Certificate2);
+            }
+
+            // If the factory is not an IMsalMtlsHttpClientFactory, use it to get a default HttpClient
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:157]: you should log a warning here"?

File: src/client/Microsoft.Identity.Client/Http/HttpManagerFactory.cs [hunk 4]
```diff
@@ -14,16 +14,19 @@ namespace Microsoft.Identity.Client.Http
     /// </summary>
     internal sealed class HttpManagerFactory
     {
-        public static IHttpManager GetHttpManager(IMsalHttpClientFactory httpClientFactory, bool withRetry, bool isManagedIdentity)
+        public static IHttpManager GetHttpManager(
+            IMsalHttpClientFactory httpClientFactory,
+            bool withRetry,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpManagerFactory.cs:19]: Can you get rid of these 2 flags? Just reading the code is tricky here. Might as well have the upper layer inject the retry policy directly.

File: src/client/Microsoft.Identity.Client/Http/HttpRetryCondition.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Net;
+
+namespace Microsoft.Identity.Client.Http
+{
+    internal static class HttpRetryConditions
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpRetryCondition.cs:8]: Have you seen @Robbie-Microsoft 's similar abstraction - https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/4d3c84b3f5b633b269e46e03382fa8d834886ab0/lib/msal-node/src/retry/LinearRetryPolicy.ts#L7

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 6]
```diff
@@ -27,123 +26,38 @@ namespace Microsoft.Identity.Client.Http
...
-        {
-            return SendRequestAsync(endpoint, headers, null, HttpMethod.Get, logger, cancellationToken: cancellationToken);
-        }
-
         /// <summary>
-        /// Performs the GET request just like <see cref="SendGetAsync(Uri, IDictionary{string, string}, ILoggerAdapter, bool, CancellationToken)"/>
-        /// but does not throw a ServiceUnavailable service exception. Instead, it returns the <see cref="HttpResponse"/> associated
-        /// with the request.
+        /// A new instance of the HTTP manager with a retry *condition*. The retry policy hardcodes: 
+        /// - the number of retries (1)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:34]: See comment below. Can you find an abstraction that eliminates these 2 hardcoded things? See Robbie's work in MSAL JS

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 7]
```diff
@@ -27,123 +26,38 @@ namespace Microsoft.Identity.Client.Http
...
-            IDictionary<string, string> headers,
+            Dictionary<string, string> headers,
             HttpContent body,
             HttpMethod method,
             ILoggerAdapter logger,
-            bool doNotThrow = false,
-            bool retry = false,
-            CancellationToken cancellationToken = default)
+            bool doNotThrow,
+            bool retry,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:53]: Why does this still have retry flag? I thught retries were injected by ctor?

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 8]
```diff
@@ -200,11 +146,39 @@ private static HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary
             return response;
         }
 
-        protected async Task<HttpResponse> ExecuteAsync(
+        private HttpClient GetHttpClient(X509Certificate2 x509Certificate2)
+        {
+            if (_httpClientFactory is IMsalMtlsHttpClientFactory msalMtlsHttpClientFactory)
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:151]: Since the default client is now an implementation of `IMsalMtlsHttpClientFactory`, this is always true but the cert is not always there. You are adding that check later in the flow. Can this be updated to check if the cert is here and then make a call to either? Also adding a check here will make it more obvious which client is used in which case.

Reason: for the service fabric cert validation, I am trying to create a client with handler. As per the discussion we had, my understanding is if the binding cert is there then we use the MtlsClient and not a default handler.

**Review by neha-bhargava — CHANGES_REQUESTED** [review]

File: src/client/Microsoft.Identity.Client/Instance/Discovery/NetworkMetadataProvider.cs [hunk 9]
```diff
@@ -34,17 +34,17 @@ internal class NetworkMetadataProvider : INetworkMetadataProvider
 
         public async Task<InstanceDiscoveryMetadataEntry> GetMetadataAsync(Uri authority, RequestContext requestContext)
         {
-            var logger = requestContext.Logger;
+            ILoggerAdapter logger = requestContext.Logger;
```
[the remark this answers is not in the document]
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Instance/Discovery/NetworkMetadataProvider.cs:37]: Using explicit types can make it a bit easier to know of what type you are dealing with when it isnt immediatly obvious. However, when the type is obvious:
```
var x = new TokenCache();
```

I think it is fine to use var. Clearly the above variable x is a `TokenCache`

File: src/client/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs [hunk 10]
```diff
@@ -29,7 +29,7 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
         {
             await ResolveAuthorityAsync().ConfigureAwait(false);
 
-            var client = new OAuth2Client(ServiceBundle.ApplicationLogger, ServiceBundle.HttpManager);
+            var client = new OAuth2Client(ServiceBundle.ApplicationLogger, ServiceBundle.HttpManager, null);
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs:32]: I see that there are a number of instances where this constructor is called with the binding cert `null`. Wondering if it makes sense to add an overload where it is assumed to be null? Or maybe a default value?

File: src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/SimpleHttpClientFactory.cs [hunk 11]
```diff
@@ -14,23 +16,73 @@ namespace Microsoft.Identity.Client.PlatformsCommon.Shared
...
+            HttpClientConfig.ConfigureRequestHeadersAndSize(httpClient);
+
+            return httpClient;
+        }
+
+        private static HttpClient CreateMtlsHttpClient(X509Certificate2 bindingCertificate)
+        {
+            CheckAndManageCache();
+
+            if (bindingCertificate == null)
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/SimpleHttpClientFactory.cs:43]: does it make sense to put this check inside the `#if SUPPORTS_MTLS` block below? If the target platform cant use MTLS the client probably wouldnt pass in a binding cert. Or would it?

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs (hunk 6 above)
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/Http/HttpManager.cs:34]: For managed identity flow we need to update the retry logic "retry policy to retry 3 times by default with a delay of 1s between retry." 
See #4662

**Comment by bgavrilMS** [comment]
Any updates on this PR @gladjohn ? Is it ready for a re-review?

**Comment by gladjohn** [comment]
> Any updates on this PR @gladjohn ? Is it ready for a re-review?

Had few test failures last week after addressing all the changes. I will start looking into this again this week. Hopefully have something out for re-review by Wednesday

File: src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityRequest.cs [hunk 12]
```diff
@@ -17,11 +17,11 @@ internal class ManagedIdentityRequest
 
         public HttpMethod Method { get; }
 
-        public IDictionary<string, string> Headers { get; }
+        public Dictionary<string, string> Headers { get; }
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityRequest.cs:20]: Why this change? Shouldn't we program to interface?

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs (hunk 1 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Http/HttpManager.cs:103]: In the line above we check if the cancellation token is cancelled and we throw. If the cancellation token is not canceled, we retry. This has always been like this.

**Review by neha-bhargava — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs (hunk 8 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Http/HttpManager.cs:147]: Discussed this with Bogdan and if the service fabric is used then server cert validation will be done and exception will be thrown in case an MTLS cert is used.

File: src/client/Microsoft.Identity.Client/Http/HttpManagerFactory.cs (hunk 4 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Http/HttpManagerFactory.cs:19]: Discussed offline, these flags cannot be eliminated but are needed to use the correct policy

File: src/client/Microsoft.Identity.Client/Http/HttpManager.cs (hunk 2 above)
Comment (neha-bhargava) [reply src/client/Microsoft.Identity.Client/Http/HttpManager.cs:136]: Will be done as part of https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4820/files. FYI @trwalke

**Review by trwalke — APPROVED** [review]
